src/Futhark/Internalise/Monomorphise.hs

```diff
@@ -243,20 +243,62 @@ transformFName loc fname t
           )
           size_args
 
+transformTypeSizes :: TypeBase Size as -> MonoM (TypeBase Size as)
+transformTypeSizes = bitraverse onDim pure
+  where
+    onDim (SizeExpr e) = SizeExpr <$> transformExp e
+    onDim (AnySize v) = pure $ AnySize v
+
+transformTypeExp :: TypeExp Info VName -> MonoM (TypeExp Info VName)
```

Comment:
I am sad that we need to do this.  TypeExps are supposed to be a non-elaborated syntactical construct that we should never need to transform.  I think the only reason we keep them around is to produce error messages for size coercions that refer to the original (non-elaborated) type entered by the programmer.  However, fixing that is a matter for a different PR.